src/SpeedyTransforms/spectral_transform.jl

```diff
@@ -340,38 +346,33 @@ function transform!(                # SPECTRAL TO GRID
+    (; lmax, mmax ) = S                             # 0-based max degree l, order m of spherical harmonics
+    (; cos_colat, sin_colat, lon_offsets ) = S  
+    (; recompute_legendre, Λ, Λs, m_truncs ) = S    # for Legendre transform
+    (; brfft_plans) = S                             # for Fourier transform
 
     @boundscheck ismatching(S, grids) || throw(DimensionMismatch(S, grids))
     @boundscheck ismatching(S, specs) || throw(DimensionMismatch(S, specs))
+    @boundscheck size(grids, 2) == size(specs, 2) <= S.nlayers || throw(DimensionMismatch(S, specs))
 
-    lmax = specs.m - 1            # 0-based maximum degree l of spherical harmonics
-    mmax = specs.n - 1            # 0-based maximum order m of spherical harmonics
-
-    # preallocate work arrays
-    gn = zeros(Complex{NF}, nfreq_max)      # phase factors for northern latitudes
-    gs = zeros(Complex{NF}, nfreq_max)      # phase factors for southern latitudes
-
-    rings = eachring(grids)                 # precomputed ring indices
+    # use scratch memory for Legendre but not yet Fourier transformed data
+    gn = S.scratch_memory_north                 # phase factors for northern latitudes
+    gs = S.scratch_memory_south                 # phase factors for northern latitudes
 
     Λw = Legendre.Work(Legendre.λlm!, Λ, Legendre.Scalar(zero(Float64)))
 
-    # loop over all specs/grids (e.g. vertical dimension)
-    # k, k_grid only differ when specs/grids have a singleton dimension
-    @inbounds for (k, k_grid) in zip(eachmatrix(specs), eachgrid(grids))
-        for j_north in 1:nlat_half              # symmetry: loop over northern latitudes only
-            j_south = nlat - j_north + 1        # southern latitude index
-            nlon = nlons[j_north]               # number of longitudes on this ring
-            nfreq  = nlon÷2 + 1                 # linear max Fourier frequency wrt to nlon
-            m_trunc = m_truncs[j_north]         # (lin/quad/cub) max frequency to shorten loop over m
-            not_equator = j_north != j_south    # is the latitude ring not on equator?
+    # INVERSE LEGENDRE TRANSFORM
+    @inbounds for k in eachmatrix(specs)        # loop over all specs/grids (e.g. vertical dimension)
+        for j in 1:nlat_half                    # symmetry: loop over northern latitudes only
+            m_trunc = m_truncs[j]               # (lin/quad/cub) max frequency to shorten loop over m
 
             # Recalculate or use precomputed Legendre polynomials Λ
-            recompute_legendre && Legendre.unsafe_legendre!(Λw, Λ, lmax, mmax, Float64(cos_colat[j_north]))
-            Λj = recompute_legendre ? Λ : Λs[j_north]
+            recompute_legendre && Legendre.unsafe_legendre!(Λw, Λ, lmax, mmax, Float64(cos_colat[j]))
+            Λj = recompute_legendre ? Λ : Λs[j]
+
+            gn[:, k, j] .= 0                    # reset scratch memory
+            gs[:, k, j] .= 0                    # reset scratch memory
```

Comment:
We could also reset the whole scratch array at once before the loop, right?

Reply:
Yes, I tried that, but this is faster ;) It must have something to do with reading into lower caches. Because I'm only setting those that I need in that iteration of the loop anyway